### DISCUSSION TITLE
Using swc as a replacement of RollUp in my project

### DISCUSSION THREAD

**Discussion by zainul-franciscus-myob** [swc-project/swc#1322] [Q&A]
Hi,

I am using RollUp to produce a bundle of React components. I used some plugins in Roll Up config. One of them is a plugin to transpile a library called Treat JS: https://github.com/seek-oss/treat. 

Is there any guide on how to write a plugin for swc ? I intend to explore writing a plugin for Treat JS.

Thank you and Happy New Year :)

**Comment by kdy1** [comment]
I'm sorry but currently plugin system is not implemented properly.
You can still write plugin, but currently you have to call `swc.transform*` manually to use plugin.

(I think it will be fixed soon.)


Currently only guide for writing plugin is https://swc.rs/blog/2019/11/30/announcing-swc-1.1.0#plugin but note that type definitions are bit outdated.

**Comment by kinthaiofficial** [comment]
Using SWC as a Rollup replacement is doable but worth understanding where the tools differ before committing:

**SWC is a transpiler/bundler, Rollup is a bundler/optimizer:**
- SWC's `spack` bundler can bundle, but it's less mature than Rollup for complex scenarios (code splitting, chunk naming, library bundling)
- SWC's transpiler (the transform part, without bundling) is essentially a 100x-faster Babel — drop-in replacement for the "transpile TypeScript/JSX" step
- Rollup excels at tree shaking, ES module output, and library authoring

**The common pattern — SWC + Rollup together:**
Use SWC as the Rollup plugin for transpilation (replaces `@rollup/plugin-babel`), but keep Rollup for the actual bundling/optimization:

```js
// rollup.config.js
import swc from 'rollup-plugin-swc3';

export default {
  input: 'src/index.ts',
  output: { file: 'dist/index.js', format: 'esm' },
  plugins: [
    swc({
      jsc: { parser: { syntax: 'typescript' }, target: 'es2020' }
    })
  ]
};
```

This gives you SWC's speed for TypeScript/JSX transformation while keeping Rollup's tree shaking and output quality.

**When to use SWC as a full Rollup replacement:**
Only if you need maximum build speed and are willing to accept SWC bundler's current limitations (less mature code splitting, fewer plugins). For library authoring where output format quality matters, Rollup is still better.

What's your specific build pipeline? That would help clarify which approach makes more sense.
